### PR TITLE
Doughnuts

### PR DESCRIPTION
**Pull request by surangap**
## Changelog
This PR will add the following functionality
- Add `pallet-doughnut` to handle doughnut transaction exectuion.
- Add `pallet_doughnut::Call::transact()` to support transaction calling with doughnuts
- Add  `pallet_doughnut::Call::revoke_doughnut()`, `pallet_doughnut::Call::revoke_holder()` to manage revocation of doughnuts
- Add `pallet_doughnut::Call:update_whitelisted_holders()` to manage whitelisted doughnut holders. 

### Dependent PRs
- https://github.com/futureversecom/trn-seed/pull/753
- https://github.com/futureversecom/trn-doughnut-rs/pull/1
- https://github.com/futureversecom/trn-pact/pull/3

### CODE REVIEW

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -2028,6 +2050,49 @@ fn validate_self_contained_inner(
...
+// 	// post dispatch
+// 	pallet_transaction_payment::ChargeTransactionPayment::<Runtime>::post_dispatch(
+// 		Some(pre),
+// 		dispatch_info,
+// 		&post_info,
+// 		len,
+// 		&res.map(|_| ()).map_err(|e| e.error),
+// 	).ok();
+// 	return Some(res)
+// }
```
Comment (JasonTulp) [inline runtime/src/lib.rs:2094]: Can remove this now?

File: pallet/common/src/lib.rs [hunk 2]
```diff
@@ -408,3 +408,13 @@ impl<T: frame_system::Config> MaintenanceCheckEVM<T> for () {
...
 }
+
+/// Generic trait to validate extrinsics satisfy some condition
+pub trait ExtrinsicChecker {
+	type Call;
+	type PermissionObject;
+	fn check_extrinsic(
+		call: &Self::Call,
+		permission_object: &Self::PermissionObject,
+	) -> DispatchResult;
```
Comment (zees-dev) [inline pallet/common/src/lib.rs:419]: Lets make `DispatchResult` an associated type; in the case of XRPL - we use `bool`.
We should default it to boolean too? - but can be overriden.

File: pallet/common/src/lib.rs [hunk 3]
```diff
@@ -408,3 +408,13 @@ impl<T: frame_system::Config> MaintenanceCheckEVM<T> for () {
 		true
 	}
 }
+
+/// Generic trait to validate extrinsics satisfy some condition
+pub trait ExtrinsicChecker {
+	type Call;
+	type PermissionObject;
```
Comment (zees-dev) [inline pallet/common/src/lib.rs:415]: Need a more generic name here; in some cases this associated type may not be used (for example for XRPL).
Maybe `Extra` makes more sense here?

Comment (zees-dev) [reply pallet/common/src/lib.rs:415]: Currently im using this for xrpl pallet:
```rust
/// Generic trait to validate extrinsics satisfy some condition
pub trait ExtrinsicChecker {
	type Call;
	type Extra = ();
	type Result = bool;
	fn check_extrinsic(call: &Self::Call, extra: &Self::Extra) -> Self::Result;
}
```

will require `#![feature(associated_type_defaults)]` at top of file (probably above the `#![cfg_attr(not(feature = "std"), no_std)]` until we upgrade our rust version.

File: runtime/src/lib.rs (hunk 1 above)
Comment (surangap) [reply runtime/src/lib.rs]: 5aebbb1

File: pallet/common/src/lib.rs (hunk 3 above)
Comment (surangap) [reply pallet/common/src/lib.rs]: 80072d6

File: e2e/test/Doughnuts.test.ts [hunk 4]
```diff
@@ -0,0 +1,1768 @@
...
+    const tip = 0;
+    const call = api.tx.balances.transfer(receiverAddress, transferAmount);
+    const version = 1;
+    const issuerPubkey = alice.publicKey;
+    const holderPubkey = holder.publicKey;
+    const feeMode = 0;
+    const expiry = 100000;
+    const notBefore = 0;
+
+    console.log("\n====  Creating Doughnut");
```
Comment (JasonTulp) [inline e2e/test/Doughnuts.test.ts:68]: nit. We should remove all these extra comments

File: pallet/doughnut/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+		<T as frame_system::Config>::Index: From<u32>,
+		<T as Config>::RuntimeCall: GetCallMetadata,
+{
+	pub fn is_self_contained(&self) -> bool {
+		matches!(self, Call::transact { .. })
+	}
+
+	pub fn check_self_contained(&self) -> Option<Result<H160, TransactionValidityError>> {
+		if let Call::transact { call, doughnut, nonce, genesis_hash, tip, signature } = self {
+			let check = || {
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:83]: nit, why use a closure for this?

File: pallet/doughnut/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+		<T as frame_system::Config>::RuntimeCall: From<<T as Config>::RuntimeCall>,
+		PostDispatchInfo: From<<<T as Config>::RuntimeCall as Dispatchable>::PostInfo>,
+		<T as frame_system::Config>::Index: From<u32>,
+		<T as Config>::RuntimeCall: GetCallMetadata,
+{
+	pub fn is_self_contained(&self) -> bool {
+		matches!(self, Call::transact { .. })
+	}
+
+	pub fn check_self_contained(&self) -> Option<Result<H160, TransactionValidityError>> {
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:81]: I feel like some of the checks in check_self_contained are redundant. i.e. We are decoding the doughnut 4 times within check_self_contained, validate_self_contained apply_self_contained and transact. Is there a way to combine validate and check? (Maybe I'm misunderstanding the purpose of the two separate methods)

Comment (JasonTulp) [reply pallet/doughnut/src/lib.rs:81]: Maybe instead we can separate out the common checks and calculations within validate and apply (i.e. getting fee_payer, genesis hash etc) into a helper function to reduce duplicate code?

File: pallet/doughnut/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+
+			// Dispatch the outer call. i.e Doughnut::transact() with None as the origin
+			let res = call.dispatch(frame_system::RawOrigin::None.into());
+			let post_info = match res {
+				Ok(info) => info,
+				Err(err) => err.post_info,
+			};
+
+			// post dispatch
+			<DoughnutFeePayerValidations<T> as SignedExtension>::post_dispatch(
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:271]: Do we need post_dispatch to be called on the sender as well?

File: pallet/doughnut/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+			+ ExtrinsicChecker<
+				Call = <Self as Config>::RuntimeCall,
+				Extra = (),
+				Result = DispatchResult,
+			>;
+		/// Weight information for the extrinsic call in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Map from collection to its public minting information
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:328]: Oops, I think this was me, but this comment is wrong 
```suggestion
	/// Map doughnut hash to blocked status
```

File: pallet/doughnut/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+		StorageMap<_, Twox64Concat, T::AccountId, bool, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config>
+	where
+		<T as frame_system::Config>::AccountId: From<H160>,
+	{
+		DoughnutCallExecuted { result: DispatchResult },
+		WhitelistedHoldersUpdated { holder: T::AccountId, enabled: bool },
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:356]: Nit. Need doc comments for events

File: pallet/doughnut/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+		TRNDomainNotfound,
+		/// Topping permissions denied.
+		ToppingPermissionDenied,
+		/// Inner call is not whitelisted
+		UnsupportedInnerCall,
+		/// Holder not whitelisted
+		HolderNotWhitelisted,
+	}
+
+	#[pallet::hooks]
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:387]: Can take this out

File: pallet/doughnut/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+		let Doughnut::V1(_) = doughnut_decoded.clone() else {
+			log!(error,"⛔️ unsupported doughnut version");
+			return Err(Error::<T>::UnsupportedDoughnutVersion)?;
+		};
+
+		Ok(doughnut_decoded)
+	}
+}
+
+/// Checks performed on a fee payer of a Doughnut transaction
```
Comment (JasonTulp) [inline pallet/doughnut/src/lib.rs:548]: Nit. Can we put this at the top of the file? Will make it easier to keep track of

File: pallet/doughnut/src/test.rs [hunk 12]
```diff
@@ -0,0 +1,1375 @@
...
+#[test]
+fn bob_to_alice_doughnut() {
+	TestExt::<Test>::default().build().execute_with(|| {
+		let issuer: TestAccount = BOB;
+		let holder: TestAccount = ALICE;
+		let doughnut =
+			make_doughnut(&holder, &issuer, FeeMode::ISSUER, "", vec![], SignatureVersion::ECDSA);
+
+		println!("issuer address (Bob): {:?}", to_hex(issuer.address().0.as_slice(), false));
+		println!("holder address (Alice): {:?}", to_hex(holder.address().0.as_slice(), false));
```
Comment (JasonTulp) [inline pallet/doughnut/src/test.rs:219]: Nit. Should probably remove these comments now

File: runtime/Cargo.toml [hunk 13]
```diff
@@ -81,6 +81,9 @@ pallet-evm-precompiles-futurepass = { path = "../evm-precompiles/futurepass", de
 pallet-evm-precompiles-futurepass-registrar = { path = "../evm-precompiles/futurepass_registrar", default-features = false }
 pallet-evm-precompiles-marketplace = { path = "../evm-precompiles/marketplace", default-features = false }
 rustc-hex = { version = "2.1.0", default-features = false }
+doughnut-rs = { git = "https://github.com/futureversecom/trn-doughnut-rs", branch = "feat/trn-doughnut", default-features = false }
+#doughnut-rs = { path = "../../../trn-doughnut-rs", default-features = false, features = ["crypto"] }
```
Comment (JasonTulp) [inline runtime/Cargo.toml:85]:
```suggestion
```

File: runtime/src/impls.rs [hunk 14]
```diff
@@ -867,6 +871,157 @@ where
...
+					..
+				}) = *inner_call_1.clone()
+				{
+					actual_call = *inner_call_2.clone();
+				} else {
+					actual_call = *inner_call_1.clone();
+				}
+			},
+			_ => actual_call = call.clone(),
+		}
```
Comment (JasonTulp) [inline runtime/src/impls.rs:902]: To remove the warning we can do something like this:
```suggestion
		let call: Self::Call = &call {
			RuntimeCall::Futurepass(pallet_futurepass::Call::proxy_extrinsic {
				call: inner_call,
				..
			}) => *inner_call.clone(),
			RuntimeCall::FeeProxy(pallet_fee_proxy::Call::call_with_fee_preferences {
				call: inner_call_1,
				..
			}) => {
				if let RuntimeCall::Futurepass(pallet_futurepass::Call::proxy_extrinsic {
					call: inner_call_2,
					..
				}) = *inner_call_1.clone()
				{
					*inner_call_2.clone();
				} else {
					*inner_call_1.clone();
				}
			},
			_ => call.clone(),
		}
```

File: runtime/src/impls.rs [hunk 15]
```diff
@@ -867,6 +871,157 @@ where
...
+
+		if pallet_maintenance_mode::MaintenanceChecker::<Runtime>::call_paused(&actual_call) {
+			return Err(frame_system::Error::<Runtime>::CallFiltered.into())
+		}
+
+		let CallMetadata { function_name, pallet_name } = actual_call.get_call_metadata();
+		// selective matching the inner call for permission validations
+		match &actual_call {
+			// Balances
+			RuntimeCall::Balances(pallet_balances::Call::transfer { dest, value }) => {
```
Comment (JasonTulp) [inline runtime/src/impls.rs:912]: Ideally we'd use AssetsExt transfer for these (to support any asset + balances) But maybe a job for the next iteration

**Review by JasonTulp — APPROVED** [review]
Lgtm! 🚀 Left a few comments, but nothing that should block merging (Can push to next update)

File: e2e/common/index.ts [hunk 16]
```diff
@@ -203,6 +203,7 @@ export const typedefs = {
 export const NATIVE_TOKEN_ID = 1;
 export const GAS_TOKEN_ID = 2;
 export const ALITH_PRIVATE_KEY = "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5bdc4da702133";
+export const ALICE_PRIVATE_KEY = "0xcb6df9de1efca7a3998a8ead4e02159d5fa99c3e0d4fd6432667390bb4726854";
```
Comment (zees-dev) [inline e2e/common/index.ts:206]: Isn't ALITH and BOB enough?

File: e2e/test/Doughnuts.test.ts
Comment (zees-dev) [inline e2e/test/Doughnuts.test.ts:1]: need to remove all console.logs - or comment them out.
Makes debugging broken test in CI more difficult

File: pallet/doughnut/src/benchmarking.rs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+	}: _(RawOrigin::None, Box::new(call), doughnut_encoded, nonce, genesis_hash, tip, signature)
+	verify {
+		// Verify success event was thrown
+		assert_last_event::<T>(Event::DoughnutCallExecuted { result: DispatchResult::Ok(()) }.into());
+	}
+
+	revoke_doughnut {
+		let alice = H160::from(hex!("e04cc55ebee1cbce552f250e85c57b70b2e2625b"));
+		let alice: T::AccountId = alice.into();
+		let doughnut_encoded = hex!("011800020a1091341fe5664bfa1782d5e04779689068c916b04cb365ec3153755684d9a10390084fdbf27d2b79d26a4f13f0ccd982cb755a661969143c37cbc49ef5b91f27000000000074726e0000000000000000000000000045000000000053797374656d00000000000000000000000000000000000000000000000000000072656d61726b000000000000000000000000000000000000000000000000000061cb11f74c449c9371e92b7e9d01c99488e106a152c5a4ff772d80d33b08418d23b000675bfad1641b58a6924382269329d115dc14c0ae5c03efd3eb36953cef00");
```
Comment (zees-dev) [inline pallet/doughnut/src/benchmarking.rs:76]: would be good to construct this if we can; so we know how the value is derived

File: pallet/doughnut/src/mock.rs [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+	fn weight_to_fee(weight: &Weight) -> Self::Balance {
+		FeeControl::length_to_fee(weight)
+	}
+}
+parameter_types! {
+	pub const OperationalFeeMultiplier: u8 = 1;
+	pub const XrpAssetId: AssetId = XRP_ASSET_ID;
+}
+pub type XrpCurrency = pallet_assets_ext::AssetCurrency<Test, XrpAssetId>;
+impl pallet_transaction_payment::Config for Test {
```
Comment (zees-dev) [inline pallet/doughnut/src/mock.rs:73]: can we not use the macro instead?

File: runtime/src/lib.rs [hunk 19]
```diff
@@ -1306,6 +1315,7 @@ construct_runtime! {
 		FeeProxy: pallet_fee_proxy = 31,
 		FeeControl: pallet_fee_control = 40,
 		Xls20: pallet_xls20 = 42,
+		Doughnut: pallet_doughnut = 48,
```
Comment (zees-dev) [inline runtime/src/lib.rs:1318]: may need to be changed due to xrpl pallet being merged first

**Review by zees-dev — APPROVED** [review]
Minor changes required; no blocker from my POV.

Note: THE XRPL pallet has been merged first; so changes/updates will be required here.
Note: We also need to merge the dependent PRs first, tag and release, then reference the released crate versions instead.

File: pallet/doughnut/src/lib.rs (hunk 5 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:83]: copied code from ethereum pallet :D

File: pallet/doughnut/src/lib.rs (hunk 6 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:81]: regarding decoding of doughnut 4 times, we don't have a workaround atm. no way to pass the decoded doughnut downstream due to trait restrictions.
But we can add a common function to include all the common validations.

File: pallet/doughnut/src/lib.rs (hunk 7 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:271]: doesn't matter. anw added for consistency

File: runtime/src/impls.rs (hunk 15 above)
Comment (surangap) [reply runtime/src/impls.rs:971]: yeah, need a discussion to select the whitelist

File: e2e/common/index.ts (hunk 16 above)
Comment (surangap) [reply e2e/common/index.ts:206]: Alice to bob test cases in early tests I think

File: pallet/doughnut/src/mock.rs (hunk 18 above)
Comment (surangap) [reply pallet/doughnut/src/mock.rs:73]: tried this, there were more dependent pallets required to use the macro

File: runtime/src/lib.rs (hunk 19 above)
Comment (surangap) [reply runtime/src/lib.rs:1337]: `48` looks good

File: e2e/test/Doughnuts.test.ts (hunk 4 above)
Comment (surangap) [reply e2e/test/Doughnuts.test.ts]: [62e90bf](https://github.com/futureversecom/trn-seed/pull/760/commits/62e90bf935de42565c9a51b52f1e6c1fd2f680a8)

File: pallet/doughnut/src/lib.rs (hunk 7 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:282]: [62e90bf](https://github.com/futureversecom/trn-seed/pull/760/commits/62e90bf935de42565c9a51b52f1e6c1fd2f680a8)

File: pallet/doughnut/src/lib.rs (hunk 8 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs]: [62e90bf](https://github.com/futureversecom/trn-seed/pull/760/commits/62e90bf935de42565c9a51b52f1e6c1fd2f680a8)

File: pallet/doughnut/src/lib.rs (hunk 9 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs]: [ffd4629](https://github.com/futureversecom/trn-seed/pull/760/commits/ffd4629851906371c80514b54a79229c864f3b63)

File: pallet/doughnut/src/lib.rs (hunk 11 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs]: [ffd4629](https://github.com/futureversecom/trn-seed/pull/760/commits/ffd4629851906371c80514b54a79229c864f3b63)

File: pallet/doughnut/src/test.rs (hunk 12 above)
Comment (surangap) [reply pallet/doughnut/src/test.rs]: [ffd4629](https://github.com/futureversecom/trn-seed/pull/760/commits/ffd4629851906371c80514b54a79229c864f3b63)

File: runtime/Cargo.toml (hunk 13 above)
Comment (surangap) [reply runtime/Cargo.toml]: [ffd4629](https://github.com/futureversecom/trn-seed/pull/760/commits/ffd4629851906371c80514b54a79229c864f3b63)

File: runtime/src/impls.rs (hunk 14 above)
Comment (surangap) [reply runtime/src/impls.rs]: [ffd4629](https://github.com/futureversecom/trn-seed/pull/760/commits/ffd4629851906371c80514b54a79229c864f3b63)

File: pallet/doughnut/src/benchmarking.rs (hunk 17 above)
Comment (surangap) [reply pallet/doughnut/src/benchmarking.rs:77]: [d233d1c](https://github.com/futureversecom/trn-seed/pull/760/commits/d233d1c9293f61988dee4afda3c0f9be12741b94)

File: e2e/test/Doughnuts.test.ts
Comment (surangap) [reply e2e/test/Doughnuts.test.ts:1]: [62e90bf](https://github.com/futureversecom/trn-seed/pull/760/commits/62e90bf935de42565c9a51b52f1e6c1fd2f680a8)

File: pallet/doughnut/src/lib.rs (hunk 6 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:92]: [5fba5ce](https://github.com/futureversecom/trn-seed/pull/760/commits/5fba5ce21a3b4bb432261eee8351ab63f65716db)

**Review by JasonTulp — APPROVED** [review]

File: pallet/doughnut/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+		StorageMap<_, Twox64Concat, T::AccountId, bool, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config>
+	where
+		<T as frame_system::Config>::AccountId: From<H160>,
+	{
+		/// Doughnut transaction executed
+		DoughnutCallExecuted { result: DispatchResult },
```
Comment (zees-dev) [inline pallet/doughnut/src/lib.rs:356]: This should contain more details I think, i.e. holder, issuer, permissions, etc.
Anything you think would be useful to offchain services; I think we'd like to track this data to know which holders are using the doughnuts.

File: pallet/doughnut/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,548 @@
...
+				true => BlockedDoughnuts::<T>::insert(doughnut_hash, true),
+				false => BlockedDoughnuts::<T>::remove(doughnut_hash),
+			}
+
+			Ok(())
+		}
+
+		/// Block a holder from executing any doughnuts from a specific issuer
+		#[pallet::weight(T::WeightInfo::revoke_holder())]
+		pub fn revoke_holder(
```
Comment (zees-dev) [inline pallet/doughnut/src/lib.rs:489]: Same here, remember to emit the event.

**Review by zees-dev — CHANGES_REQUESTED** [review]
Will approve after we get the event changes in - which seem to be quite important imo.

Otherwise looks good!

File: pallet/doughnut/src/lib.rs (hunk 20 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs]: [52cd04c](https://github.com/futureversecom/trn-seed/pull/760/commits/52cd04c6077ca696e0f3f20d269d43e84c1bace4)

File: pallet/doughnut/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,548 @@
...
+			Self::deposit_event(Event::<T>::DoughnutCallExecuted {
+				result: e.map(|_| ()).map_err(|e| e.error),
+			});
+
+			Ok(())
+		}
+
+		/// Block a specific doughnut to be used
+		#[pallet::weight(T::WeightInfo::revoke_doughnut())]
+		pub fn revoke_doughnut(
```
[the remark this answers is not in the document]
Comment (surangap) [reply pallet/doughnut/src/lib.rs:475]: [52cd04c](https://github.com/futureversecom/trn-seed/pull/760/commits/52cd04c6077ca696e0f3f20d269d43e84c1bace4)

File: pallet/doughnut/src/lib.rs (hunk 21 above)
Comment (surangap) [reply pallet/doughnut/src/lib.rs:501]: [52cd04c](https://github.com/futureversecom/trn-seed/pull/760/commits/52cd04c6077ca696e0f3f20d269d43e84c1bace4)

File: runtime/src/impls.rs [hunk 23]
```diff
@@ -1028,6 +1028,35 @@ impl seed_pallet_common::ExtrinsicChecker for DoughnutCallValidator {
 					.map_err(|_| pallet_doughnut::Error::<Runtime>::ToppingPermissionDenied)?;
 				Ok(())
 			},
+			// AssetsExt
+			RuntimeCall::AssetsExt(pallet_assets_ext::Call::transfer {
```
Comment (JasonTulp) [inline runtime/src/impls.rs:1032]: Nice :+1: This will be useful

**Review by JasonTulp — APPROVED** [review]
Looks good, cheers for adding AssetsExt::transfer + expiry validation

**Review by zees-dev — APPROVED** [review]
LGTM!

**Pull request merged**